core/src/main/java/org/mule/runtime/core/internal/routing/ChoiceRouter.java

```diff
@@ -6,51 +6,190 @@
+    return processToApply(event, this);
+  }
+
+  @Override
+  public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
+    return Flux.merge(new SinkRouter(publisher, routes).collectPublishers());
+  }
+
+  public void updateStatistics(Processor processor) {
+    if (getRouterStatistics() != null && getRouterStatistics().isEnabled()) {
+      getRouterStatistics().incrementRoutedMessage(processor);
+    }
+  }
+
+  public RouterStatistics getRouterStatistics() {
+    return routerStatistics;
+  }
+
+  @Override
+  public void setRouterStatistics(RouterStatistics routerStatistics) {
+    this.routerStatistics = routerStatistics;
+  }
+
+  @Override
+  public String toString() {
+    return format("%s [flow=%s, started=%s]", getClass().getSimpleName(), getLocation().getRootContainerName(), started);
+  }
+
+  /**
+   * Router that generates executable routes for each one configured and handles the logic of deciding which one will be executed
+   * when an event arrives. It also provides access to the routes publishers so they can be merged together.
+   */
+  private class SinkRouter {
+
+    private final Flux<CoreEvent> router;
+    private final List<ExecutableRoute> routes;
+
+    SinkRouter(Publisher<CoreEvent> publisher, List<ProcessorRoute> routes) {
+      this.routes = routes.stream().map(ProcessorRoute::toExecutableRoute).collect(toList());
+      router = from(publisher)
+          .doOnNext(checkedConsumer(this::route))
+          .doOnComplete(() -> this.routes.stream()
+              .forEach(executableRoute -> executableRoute.complete()));
+    }
+
+    /**
+     * Decides which route should execute for an incoming event and executes it, purposely separating those actions so that a
+     * single {@link ExpressionManagerSession} can be used for all routes decision process.
+     *
+     * @param event the incoming event
+     */
+    private void route(CoreEvent event) {
+      ExecutableRoute selectedRoute;
+      try (ExpressionManagerSession session = expressionManager.openSession(getLocation(), event, NULL_BINDING_CONTEXT)) {
+        selectedRoute = routes.stream().filter(route -> route.shouldExecute(session)).findFirst().get();
+      }
+      selectedRoute.execute(event);
+      updateStatistics(selectedRoute.getProcessor());
+    }
+
+    /**
+     * @return the publishers of all routes so they can be merged and subscribed, including a phantom one to guarantee the
+     *         subscription of the router occurs after all routes and with the general context.
+     */
+    private List<Flux<CoreEvent>> collectPublishers() {
+      List<Flux<CoreEvent>> routes = this.routes.stream()
+          .map(ExecutableRoute::getPublisher)
+          .collect(toCollection(ArrayList::new));
+      // Decorate last route publisher to trigger 'router' subscription
+      Integer lastRouteIndex = routes.size() - 1;
+      routes.set(lastRouteIndex, subscribeFluxOnPublisherSubscription(routes.get(lastRouteIndex), router));
```

Comment:
this complexity comes from the stream up there right? why no just use an iteration to make this more clear?

Reply:
Agree. Much clearer doing it iteratively.